### PR TITLE
Bridge usernames typed in chat correctly from Slack -> IRC

### PR DESCRIPTION
**Pull request by kkuehler**
A start to resolving #15. Regex searches a message for usernames enclosed in <@...> and replaces that substring with their username before sending the message to irc.

### CODE REVIEW

File: slackbridge/bots.py [hunk 1]
```diff
@@ -8,16 +9,17 @@
 
 
 class IRCBot(irc.IRCClient):
-    pass
+    def __init__(self, sc):
+        self.sc = sc
```
Comment (matthew-mcallister) [inline slackbridge/bots.py]: Whoever originally named this `sc` did us a disservice :P

Comment (gnowxilef) [reply slackbridge/bots.py]: easy enough to find and replace no?

File: slackbridge/bots.py [hunk 2]
```diff
@@ -91,10 +92,12 @@ def check_slack_rtm(self):
 
 class UserBot(IRCBot):
 
-    def __init__(self, nickname, realname, user_id, channels):
+    def __init__(self, nickname, realname, user_id, users, channels):
+        self.topics = {}
         self.nickname = '{}-slack'.format(utils.strip_nick(nickname))
         self.realname = realname
         self.user_id = user_id
+        self.users = users
```
Comment (jvperrin) [inline slackbridge/bots.py]: I think instead of keeping this as a local variable in each bot that's it's probably better to put this as a class variable under `IRCBot` so that the bridge bot and all the user bots have access to the same data.

File: slackbridge/bots.py [hunk 3]
```diff
@@ -108,4 +111,17 @@ def signedOn(self):
             self.join('#{}'.format(channel['name']))
 
     def post_to_irc(self, channel, message):
-        self.msg(channel, message)
+        self.msg(channel, self.__format_message(message))
+
+    def __format_message(self, message):
```
Comment (jvperrin) [inline slackbridge/bots.py]: I'd use a single underscore to signify "private" methods instead of double underscores. Apparently there [is actually a difference](https://stackoverflow.com/a/41762806/1979001) in how python treats them.

File: slackbridge/bots.py [hunk 4]
```diff
@@ -108,4 +111,17 @@ def signedOn(self):
...
+        match_ids = re.findall(r'(<\@([A-Z0-9]{9,})\>)', message)
+        # If for some weird reason someone mentions the same person
+        # twice in one message, we avoid extra api calls
```
Comment (jvperrin) [inline slackbridge/bots.py]: There aren't even any API calls here any more, so I believe this comment should be updated

File: slackbridge/bots.py [hunk 5]
```diff
@@ -108,4 +111,17 @@ def signedOn(self):
...
+        for replace, uid in set(match_ids):
+            user_info = next(
+                (user for user in self.users if user['id'] == uid), None)
```
Comment (jvperrin) [inline slackbridge/bots.py]: I feel like splitting this across lines this way is confusing, maybe do this instead?

```python
user_info = next(
    (user for user in self.users if user['id'] == uid),
    None,
)
```

File: slackbridge/bots.py [hunk 6]
```diff
@@ -108,4 +111,17 @@ def signedOn(self):
...
+            if user_info:
+                target_nick = '{}-slack'.format(
+                    utils.strip_nick(user_info['name']))
```
Comment (jvperrin) [inline slackbridge/bots.py]: Same here, I think this format would be clearer for long lines:
```python
target_nick = '{}-slack'.format(
    utils.strip_nick(user_info['name']),
)
```

File: slackbridge/factories.py [hunk 7]
```diff
@@ -53,9 +53,10 @@ def add_user_bot(self, user_bot):
 
 class UserBotFactory(BotFactory):
 
-    def __init__(self, bridge_bot_factory, slack_user, channels):
+    def __init__(self, bridge_bot_factory, slack_user, users, channels):
```
Comment (jvperrin) [inline slackbridge/factories.py]: Putting this data into the shared `IRCBot` would also eliminate the need to pass this around everywhere, would be pretty nice.

File: slackbridge/bots.py [hunk 8]
```diff
@@ -15,9 +16,9 @@ class BridgeBot(IRCBot):
     nickname = 'slack-bridge'
 
     def __init__(self, sc, nickserv_pw, slack_uid, channels, user_bots):
+        self.sc = sc
```
Comment (jvperrin) [inline slackbridge/bots.py]: I do think changing this to `slack_client` or something would be good, as @matthew-mcallister mentioned earlier, but that's my fault for picking a bad name, so it's something that would be good to do in a separate pull request sometime.

File: slackbridge/bots.py (hunk 1 above)
Comment (jvperrin) [reply slackbridge/bots.py]: Haha yeah, I should have picked a better name for this variable, `sc` is not very descriptive at all, but I was getting tired of prefixing everything with `slack_` or `slack-`.

**Comment by kkuehler** [comment]
@jvperrin The suggested changes have been made.

File: slackbridge/bots.py [hunk 9]
```diff
@@ -92,6 +93,7 @@ def check_slack_rtm(self):
 class UserBot(IRCBot):
 
     def __init__(self, nickname, realname, user_id, channels):
+        self.topics = {}
```
Comment (jvperrin) [inline slackbridge/bots.py]: Is this used at all?
